### PR TITLE
editoast : remove invalid trains from results and STDCM

### LINKED ISSUE #4699: back : STDCM : filter invalid trains
## Description 

We can only use STDCM on valid trains. 

## Acceptance Criteria 

- [ ] only valid trains should be in STDCM

### LINKED ISSUE #4700: back : Do not take into account invalid trains in /train_schedule/results
## Description 

Do not take into account invalid trains in /train_schedule/results

### PR DESCRIPTION
**Pull request by younesschrifi**
Close #4700 
Close #4699

### CODE REVIEW

File: editoast/src/models/timetable.rs [hunk 1]
```diff
@@ -149,6 +149,24 @@ impl Timetable {
...
+        use crate::tables::osrd_infra_scenario::dsl as scenario_dsl;
+        let timetable_id = self.id.unwrap();
+        block(move || {
+            let mut conn = db_pool.get().unwrap();
+            scenario_dsl::osrd_infra_scenario
+                .filter(scenario_dsl::timetable_id.eq(timetable_id))
+                .inner_join(infra_dsl::osrd_infra_infra)
+                .select(infra_dsl::version)
+                .first::<String>(&mut conn)
+                .unwrap()
```
Comment (leovalais) [inline editoast/src/models/timetable.rs]: Better use `expect` in case something breaks

File: editoast/src/views/stdcm/mod.rs [hunk 2]
```diff
@@ -232,29 +242,57 @@ async fn parse_stdcm_steps(
...
+    schedules.retain(|schedule| {
+        check_train_validity(
+            &infra_version,
+            schedule.rollingstock_version.unwrap(),
+            &schedule.infra_version.clone().unwrap(),
+            rolling_stock_version,
+        )
+        .is_empty()
+    });
+    let train_schedules = schedules.clone();
```
Comment (leovalais) [inline editoast/src/views/stdcm/mod.rs]: Can't you return the moved `schedules` in the closure below and avoid cloning?

File: editoast/src/views/train_schedule/mod.rs [hunk 3]
```diff
@@ -335,9 +335,32 @@ async fn get_results(
...
         })?;
-    let schedules = timetable.get_train_schedules(db_pool.clone()).await?;
-
-    if schedules.is_empty() {
+    let infra_version = timetable
+        .infra_version_from_timetable(db_pool.clone())
+        .await;
+    let pool = db_pool.clone();
+    let mut schedules = timetable.get_train_schedules(pool.clone()).await?;
+    let schedules = block::<_, Result<_>>(move || {
```
Comment (leovalais) [inline editoast/src/views/train_schedule/mod.rs]: Can you try to factorize this part (and the STDCM one) to keep that function smaller?

**Review by leovalais — CHANGES_REQUESTED** [review]
Read the code and left a few comments, let me know if you want to discuss those :)
I'll test your changes a bit later

File: editoast/src/models/timetable.rs [hunk 4]
```diff
@@ -149,6 +149,24 @@ impl Timetable {
...
+        use crate::tables::osrd_infra_scenario::dsl as scenario_dsl;
+        let timetable_id = self.id.unwrap();
+        block(move || {
+            let mut conn = db_pool.get().unwrap();
+            scenario_dsl::osrd_infra_scenario
+                .filter(scenario_dsl::timetable_id.eq(timetable_id))
+                .inner_join(infra_dsl::osrd_infra_infra)
+                .select(infra_dsl::version)
+                .first::<String>(&mut conn)
+                .expect("error while retrieving infra")
```
Comment (leovalais) [inline editoast/src/models/timetable.rs]:
```suggestion
                .expect("could not retrieve the version of the infra of a scenario using its timetable")
```

File: editoast/src/models/train_schedule.rs [hunk 5]
```diff
@@ -403,3 +409,30 @@ pub struct ScheduledPoint {
...
+    db_pool: Data<DbPool>,
+    timetable_id: i64,
+    infra_version: String,
+) -> Result<Vec<TrainSchedule>> {
+    let mut schedules = get_timetable_train_schedules(timetable_id, db_pool.clone()).await?;
+    block::<_, Result<_>>(move || {
+        let mut conn = db_pool.clone().get()?;
+        schedules.retain(|schedule| {
+            let rolling_stock =
+                LightRollingStockModel::retrieve_conn(&mut conn, schedule.rolling_stock_id)
```
Comment (leovalais) [inline editoast/src/models/train_schedule.rs:423]: The `block` is unnecessary, you can just `retrieve(db_pool, id)` the LightRollingStock.

File: editoast/src/views/train_schedule/mod.rs [hunk 6]
```diff
@@ -335,9 +335,12 @@ async fn get_results(
...
         })?;
-    let schedules = timetable.get_train_schedules(db_pool.clone()).await?;
-
-    if schedules.is_empty() {
+    let infra_version = timetable
+        .infra_version_from_timetable(db_pool.clone())
+        .await;
+    let schedules =
+        filter_invalid_trains(db_pool.clone(), timetable.id.unwrap(), infra_version).await?;
+    if schedules.clone().is_empty() {
```
Comment (leovalais) [inline editoast/src/views/train_schedule/mod.rs]:
```suggestion
    if schedules.is_empty() {
```

File: editoast/src/views/stdcm/mod.rs [hunk 7]
```diff
@@ -232,29 +235,50 @@ async fn parse_stdcm_steps(
...
-        .flat_map(|simulation| {
-            let sim = simulation
-                .simulation_output
-                .eco_simulation
-                .as_ref()
-                .unwrap_or(&simulation.simulation_output.base_simulation);
-            sim.route_occupancies
+        .filter_map(|simulation| {
+            if let Some(schedule) = schedules
+                .clone()
```
Comment (leovalais) [inline editoast/src/views/stdcm/mod.rs]:
```suggestion
```

You don't need to clone the whole Vec here

**Review by leovalais — CHANGES_REQUESTED** [review]
A few extra clones that I didn't spot last time, but otherwise LGTM, I'll try to test it locally today

File: editoast/src/models/train_schedule.rs (hunk 5 above)
Comment (younesschrifi) [reply editoast/src/models/train_schedule.rs:423]: the problem here is that `retain` function can't have an async block in it. That's why I used the `block`.

**Review by leovalais — APPROVED** [review]
Tested both /train_schedule/results and /stdcm using curl.
The OpenAPI needs to be updated, either change `openapi_legacy.myl` or add utoipa annotations (you can ask Auguste to show you, he's done it a bunch of times already). If you decide to use utoipa, please make a different commit to help with the review :)

**Pull request merged**